// mahabhuta/dist/CustomElement.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomElement = void 0;
const Mahafunc_1 = require("./Mahafunc");
const util = __importStar(require("util"));
const index_1 = require("./index");
/**
 * Implements an HTML-ish element that is replaced with
 * some other HTML.  For example, <embed-video> might take
 * an href= and other attributes to describe a video from
 * a known service, the process function discerns the HTML code
 * to use for the player, rendering that into the output.
 */
class CustomElement extends Mahafunc_1.Mahafunc {
    /**
     * The name of the element that is implemented by
     * this function.
     */
    get elementName() {
        throw new Error("The 'elementName' getter must be overridden");
    }
    /**
     * The selector for a _CustomElement_ implementation
     * is simply the name of the element.
     */
    get selector() { return this.elementName; }
    async process($element, metadata, setDirty, done) {
        throw new Error("The 'process' function must be overridden");
    }
    async processAll($, metadata, setDirty) {
        try {
            let elements = this.findElements($);
            if (elements.length <= 0)
                return;
            // Performance testing
            const _start = new Date();
            for (let element of elements) {
                let replaceWith = await this.process($(element), metadata, setDirty);
                // console.log(`CustomElement ${this.elementName} process returned ${replaceWith}`);
                $(element).replaceWith(replaceWith);
            }
            /* if (this.elementName === "site-verification") {
                console.log(`CustomElement ${this.elementName} `, $.html());
            } */
            // Performance testing
            (0, index_1.logPerformance)(_start, `CustomElement ${this.array.name} ${this.elementName}`);
        }
        catch (e) {
            console.error(`CustomElement ${this.elementName} Errored with ${util.inspect(e)}`);
            throw e;
        }
    }
}
exports.CustomElement = CustomElement;
