// mahabhuta/dist/Munger.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Munger = void 0;
const Mahafunc_1 = require("./Mahafunc");
const util = __importStar(require("util"));
const index_1 = require("./index");
class Munger extends Mahafunc_1.Mahafunc {
    get elementName() {
        return ''; // This is a default
    }
    /**
     * For _Munger_ classes, the `process` function has access
     * to the entire HTML for the page.  Hence, it is passed
     * both `$` for the whole page, and `$element` matching
     * a specific element on the page.
     *
     * @param $
     * @param $element
     * @param metadata
     * @param setDirty
     * @param done
     */
    async process($, $element, metadata, setDirty, done) {
        throw new Error("The 'process' function must be overridden");
    }
    async processAll($, metadata, setDirty) {
        try {
            var elements = this.findElements($);
            if (elements.length <= 0)
                return;
            // Performance testing
            const _start = new Date();
            // console.log(`Munger ${this.array.name} ${this.elementName} found ${elements.length} elements`);
            for (let element of elements) {
                await this.process($, $(element), metadata, setDirty);
            }
            // Performance testing
            (0, index_1.logPerformance)(_start, `Munger ${this.array.name} ${this.elementName}`);
        }
        catch (e) {
            console.error(`Munger ${this.selector} Errored with ${util.inspect(e)}`);
            throw e;
        }
    }
}
exports.Munger = Munger;
